## Лабораторная работа 1. Знакомство с библиотеками для анализа данных

Результат лабораторной работы − отчет в формате Jupyter notebook'а. Постарайтесь сделать ваш отчет интересным рассказом, последовательно отвечающим на вопросы из заданий. Помимо ответов на вопросы, в отчете так же должен быть код, однако чем меньше кода, тем лучше всем: мне − меньше проверять, вам —  проще найти ошибку или дополнить эксперимент. При проверке оценивается четкость ответов на вопросы, аккуратность отчета и кода.

### О задании

Лабораторная работа №1 направлена на работу с пакетами NumPy, Pandas и Matplotlib путём решения набора задач. В данном задании разрешено пользоваться только стандартной библиотекой языка Python, а также пакетами NumPy, Pandas и Matplotlib. Целью задания является редактирование данного jupyter notebook-a (в части блоков с кодом на python) таким образом, чтобы были реализованы указанные в задании функции.

### Оценивание и штрафы

* Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи)
* Максимально допустимая оценка за работу — 16 баллов
* При сдаче до soft deadline можно получить максимум 100% баллов
* При просрочке не более, чем на 5 минут — максимум 90% баллов
* При просрочке не более, чем на 1 час — максимум 80% баллов
* При сдаче до hard deadline — максиму 70% баллов
* Сдавать задание после hard deadline нельзя
* «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов и понижают карму
* Если вы нашли решение какого-то из заданий в открытом источнике, необходимо прислать ссылку на этот источник (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник)
* Не оцениваются задания с удалёнными формулировкам
* Не оценивается лабораторная работа целиком, если она была выложена в открытый источник

## 0. Jupyter notebook (0 баллов)

Сейчас мы находимся в jupyter-ноутбуке (или ipython-ноутбуке). Ноутбук состоит из ячеек, каждая из которых может быть либо ячейкой с кодом, либо ячейкой с текстом. Текст поддерживает markdown-разметку и формулы в Latex.

In [2]:
# ячейка с кодом, при выполнении которой появится output
2 + 2

4

А это ___ячейка с текстом___.

[Здесь](https://athena.brynmawr.edu/jupyter/hub/dblank/public/Jupyter%20Notebook%20Users%20Manual.ipynb) находится <s>не</s>большая заметка о используемом языке разметки Markdown. Он позволяет:

0. Составлять упорядоченные списки
1. #Делать 
##заголовки 
###разного уровня
3. Выделять *текст* <s>при</s> **необходимости**
4. Добавлять [ссылки](http://imgs.xkcd.com/comics/the_universal_label.png)


* Составлять неупорядоченные списки

Делать вставки с помощью LaTex:
    
$
\left\{
\begin{array}{ll}
x = 16 \sin^3 (t) \\ 
y = 13 \cos (t) - 5 \cos (2t) - 2 \cos (3t) - \cos (4t) \\
t \in [0, 2 \pi]
\end{array}
\right.$

А ещё можно вставлять картинки:
<img src="https://st2.depositphotos.com/1177973/9266/i/950/depositphotos_92668716-stock-photo-red-cat-with-computer-keyboard.jpg" style="width: 400px">

## 1. Табличные данные и Pandas (5 баллов)

Материалы:
1. [Pandas за 10 минут из официального руководства](http://pandas.pydata.org/pandas-docs/stable/10min.html)
2. [Документация](http://pandas.pydata.org/pandas-docs/stable/index.html) (стоит обращаться, если не понятно, как вызывать конкретный метод)
3. [Примеры использования функционала](http://nbviewer.jupyter.org/github/justmarkham/pandas-videos/blob/master/pandas.ipynb)

Многие из заданий можно выполнить несколькими способами. Не существуют единственно верного, но попробуйте максимально задействовать арсенал pandas и ориентируйтесь на простоту и понятность вашего кода.

In [3]:
import pandas as pd

Данные можно скачать [отсюда](https://yadi.sk/i/jo_DOAV4_nUQaQ).

**(0.25 балла) Задание 1.1.** Откройте файл с таблицей. Выведите последние 10 строк.

Посмотрите на данные и скажите, что они из себя представляют, сколько в таблице строк, какие столбцы?

In [4]:
data = pd.read_csv('data.csv')

In [5]:
data.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

In [6]:
data.tail(10)

order_id  quantity            item_name  \
4612      1831         1        Carnitas Bowl   
4613      1831         1                Chips   
4614      1831         1        Bottled Water   
4615      1832         1   Chicken Soft Tacos   
4616      1832         1  Chips and Guacamole   
4617      1833         1        Steak Burrito   
4618      1833         1        Steak Burrito   
4619      1834         1   Chicken Salad Bowl   
4620      1834         1   Chicken Salad Bowl   
4621      1834         1   Chicken Salad Bowl   

                                     choice_description item_price  
4612  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $9.25   
4613                                                NaN     $2.15   
4614                                                NaN     $1.50   
4615   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]     $8.75   
4616                                                NaN     $4.45   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75

In [7]:
print(f'Row amount: {data.shape[0]}')
print(f'Col amount: {data.shape[1]}')

Row amount: 4622
Col amount: 5


In [8]:
data.dtypes

order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

Данные представляют из себя матрицу размера 4622х5.

**(0.25 балла) Задание 1.2.** Ответьте на вопросы:
1. Сколько заказов попало в выборку?
2. Сколько уникальных категорий товара было куплено? (item_name)

In [9]:
# 1
print(f'Number of orders: {data["order_id"].nunique()}')

Number of orders: 1834


In [10]:
print(f'Number of unique items: {data["item_name"].nunique()}')

Number of unique items: 50


**(0.25 балла) Задание 1.3.** Есть ли в данных пропуски? В каких колонках? 

In [11]:
print(f'{data.isna().sum()}\n')
print(f'There are missing values in "choice_description" column.')

order_id                 0
quantity                 0
item_name                0
choice_description    1246
item_price               0
dtype: int64

There are missing values in "choice_description" column.


Заполните пропуски пустой строкой для строковых колонок и нулём для числовых.

In [12]:
data = data.fillna(value={'choice_description': ""})
print(f'{data.isna().sum()}\n')

order_id              0
quantity              0
item_name             0
choice_description    0
item_price            0
dtype: int64



**(0.5 балла) Задание 1.4.** Посмотрите внимательнее на колонку с ценой товара. Какого она типа? Создайте новую колонку так, чтобы в ней цена была числом.

Для этого попробуйте применить функцию-преобразование к каждой строке вашей таблицы (для этого есть соответствующая функция).

In [13]:
def convert_price_to_float(value):
    value = value[1:]
    return float(value)

In [14]:
print(f'Column type: {data["item_price"].dtype}')
data['price'] = data['item_price'].apply(convert_price_to_float)
print(f'New column type: {data["price"].dtype}')

Column type: object
New column type: float64


In [15]:
data.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  price  
0                                                        $2.39    2.39  
1                                       [Clementine]     $3.39    3.39  
2                                            [Apple]     $3.39    3.39  
3                                                        $2.39    2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   16.98

Какая средняя/минимальная/максимальная цена у товара? 

In [16]:
print(f'Min price: {data["price"].min()}')
print(f'Mean price: {data["price"].mean():.2f}')

Min price: 1.09
Mean price: 7.46


Удалите старую колонку с ценой.

In [17]:
data = data.drop('item_price', axis=1)

In [18]:
data.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  price  
0                                                      2.39  
1                                       [Clementine]   3.39  
2                                            [Apple]   3.39  
3                                                      2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...  16.98

**(0.25 балла) Задание 1.5.** Какие 5 товаров были самыми дешёвыми и самыми дорогими? (по item_name)

Для этого будет удобно избавиться от дубликатов и отсортировать товары. Не забудьте про количество товара.

In [19]:
items = data.drop_duplicates('item_name')[['item_name', 'price']]
items = items.sort_values(by='price')

In [20]:
print('Top-5 the cheapest')
print(items.head())

Top-5 the cheapest
             item_name  price
34       Bottled Water   1.09
263  Canned Soft Drink   1.25
6        Side of Chips   1.69
40               Chips   2.15
18         Canned Soda   2.18


In [21]:
print('Top-5 the most expensive')
print(items.tail())

Top-5 the most expensive
                item_name  price
1132  Carnitas Salad Bowl  11.89
1229  Barbacoa Salad Bowl  11.89
4            Chicken Bowl  16.98
445                  Bowl  22.20
281      Steak Salad Bowl  23.78


**(0.25 балла) Задание 1.6.** Сколько раз клиенты покупали больше 1 Chicken Bowl (item_name)?

In [22]:
sum(data[data['item_name'] == 'Chicken Bowl']['quantity'] > 1)

33

**(0.5 балла) Задание 1.7.** Какой средний чек у заказа? Сколько в среднем товаров покупают?

Если необходимо провести вычисления в терминах заказов, то будет удобно сгруппировать строки по заказам и посчитать необходимые статистики.

In [25]:
data_grouped = data[['order_id', 'quantity', 'price']].groupby(by='order_id').sum()

In [27]:
print(f'Mean amount of items per order: {data_grouped.mean()["quantity"]:.2f}')
print(f'Mean price per order: {data_grouped.mean()["price"]:.2f}')

Mean amount of items per order: 2.71
Mean price per order: 18.81


**(0.25 балла) Задание 1.8.** Сколько заказов содержали ровно 1 товар?

In [29]:
sum(data_grouped['quantity'] == 1)

56

**(0.25 балла) Задание 1.9.** Какая самая популярная категория товара? 

In [47]:
data[['item_name', 'quantity']].groupby('item_name').count().sort_values(by='quantity', ascending=False).iloc[0]

quantity    726
Name: Chicken Bowl, dtype: int64

**(0.5 балла) Задание 1.10.** Какие виды Burrito существуют? Какой из них чаще всего покупают? Какой из них самый дорогой? 

In [70]:
burrito_data = data[data['item_name'].str.contains('Burrito')]
print(f'The most popular item: "{burrito_data.groupby(by="item_name")["quantity"].sum().idxmax()}"')
print(f'The most expensive item: "{burrito_data.groupby(by="item_name")["price"].max().idxmax()}"')

The most popular item: "Chicken Burrito"
The most expensive item: "Chicken Burrito"


**(0.75 балла) Задание 1.11.** В каком количестве заказов есть товар, который стоит более 40% от суммы всего чека?

Возможно, будет удобно посчитать отдельно среднюю стоимость заказа, добавить ее в исходные данные и сделать необходимые проверки.

In [78]:
order_price = data[['order_id', 'price']].groupby('order_id').sum()
order_price.columns = ['order_price']

In [81]:
data.join(order_price) # join by column and index

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description  price  order_price  
0                                                         2.39          NaN  
1                                          [Clementine]   3.39        11.56  
2                                               [Apple]   3.39        16.98  
3                                                         2.39        12.67  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...  16.98        21.00  
...                                                 ...    ...          ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...  11.75          NaN  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...  11.75          NaN  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...  11.25          NaN  
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...   8.75          NaN  
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...   8.75          NaN  

[4622 rows x 6 columns]

**(0.75 балла) Задание 1.12.** Предположим, что в данных была ошибка и Diet Coke (choice_description), который стоил $1.25, должен был стоить 1.35. Скорректируйте данные в таблицы и посчитайте, на какой процент больше денег было заработано с этого товара. Не забывайте, что количество товара не всегда равно 1.

In [ ]:
# Your code here

**(0.25 балла) Задание 1.13.** Создайте новый DateFrame из матрицы, созданной ниже. Назовите колонки index, column1, column2 и сделайте первую колонку индексом.

In [ ]:
import numpy as np

data = np.random.rand(10, 3)
df = pd.DataFrame(data, columns=['index', 'column1', 'column2'], index='index')
df

Сохраните DataFrame на диск в формате csv без индексов и названий столбцов.

In [ ]:
# Your code here

## 2. Векторные вычисления и Numpy (8 баллов)

Во всех заданиях данного раздела запрещено использовать циклы (ключевые слова *for* и *while*), [list comprehension](https://habrahabr.ru/post/30232/), [map и т.п.](http://pythonicway.com/python-functinal-programming)

Для каждой задачи приведите примеры использования реализованных функций. **Без примеров баллы за задание будут снижены в 2 раза, примеры должны отличаться от тех что приведены в самом задании**.

Под матрицей в заданиях понимается двумерный [numpy.array](https://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html)

In [ ]:
import numpy as np

**(0.25 балла) Задание 2.1.** Реализуйте функцию, принимающую на вход матрицу ```X``` и два массива индексов ```indices1``` и ```indices2``` одинаковой длины и возвращающую np.array, состоящий из последовательности элементов 

```[X[indices1[0], indices2[0]], ... , X[indices1[N-1], indices2[N-1]]]```

In [120]:
def construct_array(X, indices1, indices2):
    return np.array(X[indices1, indices2])

In [122]:
ind1 = np.array([1, 2, 2])
ind2 = np.array([0, 1, 2])
X = np.array([[1, 2, 3],
             [2, 3, 4],
             [3, 4, 5]])
print(construct_array(X, ind1, ind2))

[2 4 5]


**(0.25 балла) Задача 2.2.** Реализуйте функцию, которая на вход принимает два массива ```X``` и ```Y``` (массивы могут быть n-мерные, но их размерности должны совпадать), а возвращает **True** если они равны и **False** - иначе.

In [90]:
def detect_identic(X, Y):
    if X.shape != Y.shape:
        return False
    
    if (X == Y).all():
        return True
    return False

In [91]:
X = np.eye(2)
Y = np.eye(2)
print(detect_identic(X, Y))

True


In [92]:
Z = np.ones((2, 2))
print(detect_identic(X, Z))

False


In [93]:
A = np.zeros((2, 3))
print(detect_identic(X, A))

False


**(0.25 балла) Задание 2.3.** Реализуйте функцию, которая на вход принимает цветное изображение ```X``` (трехмерный вектор размера (n, m, 3)) и возвращает среднее значение по трём каналам (вектор длины 3).

In [109]:
def mean_channel(X):
    return X.mean(axis=0).mean(axis=0)

In [123]:
arr = np.array([[[1., 1., -1.],
        [2., 1., 1.]],

       [[3., 1., -1.],
        [4., 1., 1.]],

       [[5., 1., -1.],
        [6., 1., 1.]],

       [[7., 1., -1.],
        [8., 1., 1.]],

       [[9., 1., -1.],
        [10., 1., 1.]]])
print(mean_channel(arr))

[5.5 1.  0. ]


**(0.25 балла) Задание 2.4.** Реализуйте функцию, принимающую на вход матрицу ```X``` и некоторое число ```a``` и возвращающую ближайший к числу элемент матрицы.

Например, для ```X = np.arange(0,10).reshape((2, 5))``` и ```a = 3.6``` ответом будет 4.

In [132]:
def nearest_value(X, a):
    ind = np.argmin(np.abs(X-a))
    return X.flat[ind]

In [133]:
X = np.arange(0,10).reshape((2, 5))
a = 3.6
print(nearest_value(X, a))

4


In [134]:
X = np.ones((3, 3))
print(nearest_value(X, -10))

1.0


In [140]:
X = np.random.randint(3, 10, (3, 3))
print(X)
print(nearest_value(X, 7.5))

[[7 4 3]
 [9 6 7]
 [7 6 9]]
7


**(0.5 балла) Задача 2.5.** Реализуйте функцию, принимающую на вход два одномерных массива ```x``` и ```y``` и возвращающую матрицу, в которой первый массив соответствует первому столбцу матрицы, второй - второму.

В этом задании **запрещается** пользоваться операцией транспонирования и рекомедуется воспользоваться методом [reshape](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.reshape.html).

In [205]:
def construct_matrix(x, y):
    return np.append(x, y).reshape(-1, 2, order='F')

In [208]:
print(construct_matrix([1, 2, 3], [4, 5, 6]))

[[1 4]
 [2 5]
 [3 6]]


**(0.5 балла) Задание 2.6.** Реализуйте функцию, которая на вход принимает вектор ```x```, содержащий целые числа, кроме 0, а возвращает вектор со вставленными нулями между числами.

Например, из ```[1, -5, 3]``` получить ```[1, 0, -5, 0, 3]```

In [185]:
def add_zeros(x):
    if not x.size:
        return np.array([])
    return np.append(x, [0] * x.size).reshape(2,-1).T.flatten()[:-1]

In [186]:
print(add_zeros(np.array([1, -5, 3])))
print(add_zeros(np.array([1, 2])))
print(add_zeros(np.array([1])))
print(add_zeros(np.array([])))

[ 1  0 -5  0  3]
[1 0 2]
[1]
[]


**(0.75 балла) Задание 2.7.** Реализуйте функцию для подсчёта произведения ненулевых элементов на диагонали прямоугольной матрицы

Например, для ```X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])``` ответом является 3. Если ненулевых элементов нет, функция должна возвращать ```None```.

In [221]:
def nonzero_product(X):
    diag = np.diag(X)
    non_zero = diag[diag != 0]
    if not non_zero.size:
        return None
    
    return np.prod(non_zero)

In [223]:
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
print(nonzero_product(X))
print(nonzero_product(np.array([])))
X = np.array([[-1, 0, 1, 2], [2, 0, 2, 1], [3, 0, 3, 4], [4, 4, 4, 0]])
print(nonzero_product(X))

3
None
-3


**(0.75 балла) Задание 2.8.** Реализуйте функцию, возвращающую максимальный элемент в массиве ```X``` среди элементов, перед которыми стоит нулевой.

Например, для ```X = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])``` ответом является 5. Если подходящих элементов нет, функция должна возвращать ```None```.

In [251]:
def max_element(X):
    if not X.size or not X[X == 0].size:
        return None
    
    zeros = X==0
    zeros = np.append(False, zeros[:-1])
    candidates = X[zeros]
    return np.max(candidates)

In [252]:
print(max_element(np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])))
print(max_element(np.array([0, 0, 0, 0])))
print(max_element(np.array([])))
print(max_element(np.array([6, 2, 3, 5, 7])))

5
0
None
None


**(0.75 балла) Задание 2.9.** Реализуйте функцию, принимающую на вход матрицу ```X``` и возвращающую все её уникальные строки в виде матрицы.

In [ ]:
def get_unique_rows(X):
    # Your code here

**(0.75 балла) Задача 2.10.** Реализуйте функцию, каторая во входной вещественной матрице ```X``` находит все значения ```nan``` и заменяет их на среднее арифметическое всех остальных элементов. Если все элементы матрицы ```nan```, то верните нулевую матрицу той же размерности.

In [ ]:
def replace_nans(X):
    # Your code here

**(0.75 балла) Задача 2.11.** Напишите функцию, генерирующую [матрицу Вандермонда](https://ru.wikipedia.org/wiki/Определитель_Вандермонда), принемающую на вход вектор $(x_1, ... , x_n)$.

В этом задании **запрещается** пользоваться готовыми реализациями (например, [numpy.vander](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.vander.html)), а также [np.repeat](https://docs.scipy.org/doc/numpy/reference/generated/numpy.repeat.html) и [np.transpose](https://docs.scipy.org/doc/numpy/reference/generated/numpy.transpose.html).

При решение задействуйте [np.reshape](https://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html) и/или [np.newaxis](https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.indexing.html).

In [ ]:
def vander(x):
    # Your code here

**(0.75 балла) Задача 2.12.** Даны две вещественные матрицы ```X``` и ```Y``` с одинаковым числом столбцов и, в общем случае, различным числом строк. Необходимо реализовать функцию, вычисляющую матрицу попарных [косинусных коэффициентов](https://en.wikipedia.org/wiki/Cosine_similarity) между всеми $X_i и Y_j$, где $X_i$ - $i$-ая строка матрицы $X$, а $Y_j$ - $j$-ая строка матрицы $Y$. 

В этом задании **запрещается** пользоваться готовыми реализациями, а также [np.repeat](https://docs.scipy.org/doc/numpy/reference/generated/numpy.repeat.html) и [np.transpose](https://docs.scipy.org/doc/numpy/reference/generated/numpy.transpose.html).

При решение задействуйте [np.reshape](https://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html) и/или [np.newaxis](https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.indexing.html), [np.sqrt](https://docs.scipy.org/doc/numpy/reference/generated/numpy.sqrt.html), [np.sum](https://docs.scipy.org/doc/numpy/reference/generated/numpy.sum.html) и [np.power](https://docs.scipy.org/doc/numpy/reference/generated/numpy.power.html).

In [ ]:
def count_cosine_similarity(X, Y):
    # Your code here

**(0.75 балла) Задача 2.13.** Написать функцию, которая получает на вход матрицу и масштабирует каждый её столбец, а именно вычитает из столбца его среднее значение и делит столбец на стандартное отклонение.

Для тестирования можно сгенерировать с помощью метода [numpy.random.randint](https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.randint.html) случайную матрицу и проверить на ней работу метода.

Убедитесь, что в функции не будет происходить деления на ноль, если происходит деление на ноль, то верните **None**.

In [ ]:
def scale(X):
    # Your code here

**(0.75 балла) Задача 2.14.** Пусть $N=1000$. Повторите $N$ раз следующий эксперимент: сгенерируйте две матрицы размера $N×N$ из стандартного нормального распределения, перемножьте их (как матрицы) и найдите максимальный элемент. Какое среднее значение по экспериментам у максимальных элементов? 95-процентная квантиль? 

При решение задачи для повторения экспериментов воспользуйтесь [list comprehension](https://habrahabr.ru/post/30232/), а также [tqdm_notebook](https://habrahabr.ru/post/276725/) - для отслеживания прогресса.

In [ ]:
# Your code here

## 3. Визуализации и Matplotlib (3 балла)

При работе с данными часто неудобно делать какие-то выводы, если смотреть на таблицу и числа в частности, поэтому важно уметь визуализировать данные. В этом разделе мы этим и займёмся.

У matplotlib, конечно, же есть [документация](https://matplotlib.org/users/index.html) с большим количеством [примеров](https://matplotlib.org/examples/), но для начала достаточно знать про несколько основных типов графиков:
- plot — обычный поточечный график, которым можно изображать кривые или отдельные точки;
- hist — гистограмма, показывающая распределение некоторое величины;
- scatter — график, показывающий взаимосвязь двух величин;
- bar — столбцовый график, показывающий взаимосвязь количественной величины от категориальной.

В этом задании вы попробуете построить каждый из них. Не менее важно усвоить базовые принципы визуализаций:
- на графиках должны быть подписаны оси;
- у визуализации должно быть название;
- если изображено несколько графиков, то необходима поясняющая легенда;
- все линии на графиках должны быть чётко видны (нет похожих цветов или цветов, сливающихся с фоном);
- если отображена величина, имеющая очевидный диапазон значений (например, проценты могут быть от 0 до 100), то желательно масштабировать ось на весь диапазон значений (исключением является случай, когда вам необходимо показать малое отличие, которое незаметно в таких масштабах).
- сетка на графике помогает оценить значения в точках на глаз, это обычно полезно, поэтому лучше ее отрисовывать.

**Невыполнение принципов, описанных выше, ведёт к потере баллов!**

In [113]:
import matplotlib.pyplot as plt

Работать мы будем с той же выборкой покупкок, что и в части с Pandas. Добавим новую колонку с датой покупки.

In [114]:
import datetime

start = datetime.datetime(2018, 1, 1)
end = datetime.datetime(2018, 1, 31)
delta_seconds = int((end - start).total_seconds())

dates = pd.DataFrame(index=df.order_id.unique())
dates['date'] = [
    (start + datetime.timedelta(seconds=random.randint(0, delta_seconds))).strftime('%Y-%m-%d')
    for _ in range(df.order_id.nunique())]

# если DataFrame с покупками из прошлого заказа называется не df, замените на ваше название ниже
df['date'] = df.order_id.map(dates['date'])

AttributeError: 'DataFrame' object has no attribute 'order_id'

**(0.75 балла) Задача 3.1.** Постройте гистограмму распределения сумм покупок и гистограмму средних цен отдельных видов продуктов item_name. 

Изображайте на двух соседних графиках. Для этого может быть полезен subplot.

In [ ]:
# Your code here

**(0.75 балла) Задача 3.2.** Постройте график зависимости суммы покупок от дней.

In [ ]:
# Your code here

**(0.75 балла) Задача 3.3.** Постройте средних сумм покупок по дням недели (bar plot).

In [ ]:
# Your code here

**(0.75 балла) Задача 3.4.** Постройте график зависимости денег за товар от купленного количества (scatter plot).

In [ ]:
# Your code here

Сохраните график в формате pdf (так он останется векторизованным).

In [ ]:
# Your code here